javatests/dagger/functional/assisted/AssistedFactoryTest.java

```diff
@@ -0,0 +1,302 @@
+  static class SomeEntryPoint {
+    private final SimpleFoo.Factory nestedSimpleFooFactory;
+    private final SimpleFooFactory nonNestedSimpleFooFactory;
+    private final ExtendedSimpleFooFactory extendedSimpleFooFactory;
+    private final FooFactory fooFactory;
+    private final AbstractFooFactory abstractFooFactory;
+    private final NoAssistedParametersFooFactory noAssistedParametersFooFactory;
+
+    @Inject
+    SomeEntryPoint(
+        SimpleFoo.Factory nestedSimpleFooFactory,
+        SimpleFooFactory nonNestedSimpleFooFactory,
+        ExtendedSimpleFooFactory extendedSimpleFooFactory,
+        FooFactory fooFactory,
+        AbstractFooFactory abstractFooFactory,
+        NoAssistedParametersFooFactory noAssistedParametersFooFactory) {
+      this.nestedSimpleFooFactory = nestedSimpleFooFactory;
+      this.nonNestedSimpleFooFactory = nonNestedSimpleFooFactory;
+      this.extendedSimpleFooFactory = extendedSimpleFooFactory;
+      this.fooFactory = fooFactory;
+      this.abstractFooFactory = abstractFooFactory;
+      this.noAssistedParametersFooFactory = noAssistedParametersFooFactory;
+    }
+  }
+
+  static final class Dep1 {
+    @Inject
+    Dep1(Dep2 dep2, Dep3 dep3) {}
+  }
+
+  static final class Dep2 {
+    @Inject
+    Dep2(Dep3 dep3) {}
+  }
+
+  static final class Dep3 {
+    @Inject
+    Dep3(Dep4 dep4) {}
+  }
+
+  static final class Dep4 {
+    @Inject
+    Dep4() {}
+  }
+
+  // A base interface to test that factories can reference subclasses of the assisted parameter.
+  interface AssistedDep {}
+
+  static final class AssistedDep1 implements AssistedDep {}
+
+  static final class AssistedDep2 implements AssistedDep {}
+
+  static final class SimpleFoo {
+    private final AssistedDep assistedDep;
+
+    @AssistedInject
+    SimpleFoo(@Assisted AssistedDep assistedDep) {
+      this.assistedDep = assistedDep;
+    }
+
+    @AssistedFactory
+    interface Factory {
+      // Use different parameter names than Foo to make sure we're not assuming they're the same.
+      SimpleFoo createSimpleFoo(AssistedDep factoryAssistedDep);
+
+      // A no-op method to test static methods in assisted factories
+      static void staticMethod() {
+        return;
+      }
+
+      // A no-op method to test default methods in assisted factories
+      default void defaultMethod() {
+        return;
+      }
+    }
+  }
+
+  @AssistedFactory
+  interface SimpleFooFactory {
+    // Use different parameter names than Foo to make sure we're not assuming they're the same.
+    SimpleFoo createSimpleFoo(AssistedDep1 factoryAssistedDep1);
+
+    // A no-op method to test static methods in assisted factories
+    static void staticMethod() {
+      return;
+    }
+
+    // A no-op method to test default methods in assisted factories
+    default void defaultMethod() {
+      return;
+    }
+  }
+
+  @AssistedFactory
+  interface ExtendedSimpleFooFactory extends SimpleFooFactory {}
```

Comment:
Unless I missed it, there's no test for a covariant return interface extension.
```java
interface BaseFactory {
  Base create(String name);
}

class Concrete extends Base {
  @AssistedInject Concrete(@Assisted String name, Whevs whevs) {}
}

@AssistedFactory interface ConcreteFactory extends BaseFactory {
  Concrete create(String name);
}
```
This is essential for allowing the case where you then `@Binds @IntoSet` of `Set<BaseFactory>`.

Reply:
I think you're looking for this: https://github.com/google/dagger/blob/master/javatests/dagger/functional/assisted/AssistedFactoryBindsTest.java#L46-L67

Reply:
Ah, there it is. Thanks!